### PR TITLE
[Modules] DevTest Lab Module (API Version: 2018-09-15)

### LINKED ISSUE #2420: [Feature Request]: Azure DevTest Labs (DTL) Module
### Description

Review and enable a DTL CARML module into the library:

https://learn.microsoft.com/en-us/azure/templates/microsoft.devtestlab/labs?pivots=deployment-language-bicep

### PR DESCRIPTION
**Pull request by ahmadabdalla**
# Description

Closes #2420 

This PR adds the `Microsoft.DevTestLab/Labs` module. It uses the currently published API Version `2018-09-15`. 

---
**Note**: This PR is to address the publicly documented API version `2018-09-15`. Another PR will be raised to address the undocumented YET API Version `2018-10-15-preview`. The newer API has much more features, but static tests are failing as this API version is not showing in the output to get the list of published APIs for this resource type. The module has been uplifted to support the new API in another branch [referenced here](https://github.com/Azure/ResourceModules/tree/users/ahmad/2420_dtl/modules/Microsoft.DevTestLab/labs) but cannot be merged yet.

---

## Module Parents and Children
- Labs [Parent]
  - Artifact Sources [Child]
  - Notification Channels [Child]
  - Policy Sets [Child] = 'default', doesn't use a template reference
    - Policies [Child]
  - Schedules [Child]
  - Virtual Networks [Child]

## Test Cases
 - **Min**: Only the `name` parameter.
 - **Common**: Most parameters except two (Mandatory Artifacts for Windows and Linux), which can only be used as an update operation given a dependency on the parent resource ID.

## Extension Resources
 - Role Assignments: Updated to cover roles needed for the DevTest Lab resource type.

## Bicep Warnings
 - `identity` property. Supported but not available in the documentation, but available when deployed via the Portal Template export.
 - `managedIdentities`. Supported but not available in the documentation.

## Pipeline references

As this is a new module pipeline, created in the Fork and referenced a deployment from the fork.

| Pipeline |
| - |
| [DevTestLab: Labs](https://github.com/ahmadabdalla/ResourceModules/actions/workflows/ms.devtestlab.labs.yml)|

# Type of Change

- [X] New feature (non-breaking change which adds functionality)

# Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [x] My corresponding pipelines / checks run clean and green without any errors or warnings
- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (readme)
- [x] I did format my code

### CODE REVIEW

File: modules/Microsoft.DevTestLab/labs/.test/common/dependencies.bicep [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    location: location
+}
+
+resource virtualNetwork 'Microsoft.Network/virtualNetworks@2022-01-01' = {
+    name: virtualNetworkName
+    location: location
+    properties: {
+        addressSpace: {
+            addressPrefixes: [
+                '10.0.0.0/24'
```
Comment (AlexanderSehr) [inline modules/Microsoft.DevTestLab/labs/.test/common/dependencies.bicep:21]: Shortly before the holidays there was a meeting to discuss the structure of the test. One outcome was to align all VNET deployments to use a variable `addressPrefix = '10.0.0.0/24'` as the subnet almost always shares the address space size with its vnet

Comment (AlexanderSehr) [reply modules/Microsoft.DevTestLab/labs/.test/common/dependencies.bicep:21]: I have another (draft) PR open to align all other modules

File: modules/Microsoft.DevTestLab/labs/.test/common/deploy.test.bicep [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+// ========== //
+@description('Optional. The name of the resource group to deploy for testing purposes.')
+@maxLength(90)
+param resourceGroupName string = 'ms.devtestlab.labs-${serviceShort}-rg'
+
+@description('Optional. The location to deploy resources to.')
+param location string = deployment().location
+
+@description('Optional. A short identifier for the kind of deployment. Should be kept short to not run into resource-name length-constraints.')
+param serviceShort string = 'dtlcom'
```
Comment (AlexanderSehr) [inline modules/Microsoft.DevTestLab/labs/.test/common/deploy.test.bicep:14]:
```suggestion
param serviceShort string = 'dtllcom'
```
I guess 😄 Beware, requires regeneration of the readme's

File: modules/Microsoft.DevTestLab/labs/.test/min/deploy.test.bicep [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+// ========== //
+@description('Optional. The name of the resource group to deploy for testing purposes.')
+@maxLength(90)
+param resourceGroupName string = 'ms.devtestlab.labs-${serviceShort}-rg'
+
+@description('Optional. The location to deploy resources to.')
+param location string = deployment().location
+
+@description('Optional. A short identifier for the kind of deployment. Should be kept short to not run into resource-name length-constraints.')
+param serviceShort string = 'dtlmin'
```
Comment (AlexanderSehr) [inline modules/Microsoft.DevTestLab/labs/.test/min/deploy.test.bicep:14]:
```suggestion
param serviceShort string = 'dtllmin'
```
Like in the other file

File: modules/Microsoft.DevTestLab/labs/artifactSources/deploy.bicep [hunk 4]
```diff
@@ -0,0 +1,87 @@
+@sys.description('Conditional. The name of the parent lab. Required if the template is used in a standalone deployment.')
+param labName string
```
Comment (AlexanderSehr) [inline modules/Microsoft.DevTestLab/labs/artifactSources/deploy.bicep:2]: It there a reason it's using the term "parent lab" in the description? Is the "artifactSource" an old term?

File: modules/Microsoft.DevTestLab/labs/.test/common/dependencies.bicep (hunk 1 above)
Comment (ahmadabdalla) [reply modules/Microsoft.DevTestLab/labs/.test/common/dependencies.bicep:21]: @AlexanderSehr sounds good. Will look at your PR and apply it here

Comment (ahmadabdalla) [reply modules/Microsoft.DevTestLab/labs/.test/common/dependencies.bicep]: @AlexanderSehr Updated to reflect the pattern used in your PR

File: modules/Microsoft.DevTestLab/labs/.test/common/deploy.test.bicep (hunk 2 above)
Comment (ahmadabdalla) [reply modules/Microsoft.DevTestLab/labs/.test/common/deploy.test.bicep]: Fixed

File: modules/Microsoft.DevTestLab/labs/.test/min/deploy.test.bicep (hunk 3 above)
Comment (ahmadabdalla) [reply modules/Microsoft.DevTestLab/labs/.test/min/deploy.test.bicep]: Fixed

**Review by AlexanderSehr — APPROVED** [review]

**Pull request merged**